lib/Sema/CSSimplify.cpp

```diff
@@ -4960,6 +4960,29 @@ ConstraintSystem::addKeyPathApplicationConstraint(Type keypath,
                                               Type root, Type value,
                                               ConstraintLocatorBuilder locator,
                                               bool isFavored) {
+  // If this is a subscript with a KeyPath expression, add a constraint that
+  // connects the subscript's root type to the root type of the KeyPath.
+  SmallVector<LocatorPathElt, 4> path;
+  Expr *anchor = locator.getLocatorParts(path);
+  if (auto subscript = dyn_cast_or_null<SubscriptExpr>(anchor)) {
+    if (path.size() == 1 &&
+        path[0].getKind() == ConstraintLocator::SubscriptMember) {
+      if (auto indexTuple = dyn_cast<TupleExpr>(subscript->getIndex())) {
+        if (indexTuple->getNumElements() == 1) {
+          if (auto keyPathExpr = dyn_cast<KeyPathExpr>(indexTuple->getElement(0))) {
+            for (auto constraint : InactiveConstraints) {
```

Comment:
Rather than iterating over all constraints here, and checking the anchor expression, you should be able to do something like:
```
  auto *typevar = cs.getType(keyPathExpr)->getAs<TypeVariableType>();
  if (typevar) {
    CG.gatherConstraints(typevar, constraints, ConstraintGraph::GatheringKind::EquivalenceClass);
    // iterate over only those looking for a ConstraintKind::KeyPath where the constraint.getFirstType()->isEqual(typevar)
  }
```